### PR TITLE
Fix unsoundness in ByteSlice

### LINKED ISSUE #67: Unsoundness in ByteSlice::fetch
The unsafe code is here: <https://github.com/datenlord/datenlord/blob/68b29885de97ce41cacb0284162a36ffcf8dc946/async_fuse/src/fuse_request.rs#L43>

An exploit:

```rust
use std::mem;
use anyhow::Context;

pub(crate) struct ByteSlice<'a> {
    data: &'a [u8],
}

impl<'a> ByteSlice<'a>{
    pub fn new(data: &'a [u8]) -> ByteSlice<'a> {
        ByteSlice { data }
    }

    pub fn fetch_bytes(&mut self, amt: usize) -> anyhow::Result<&'a [u8]> {
        if amt > self.data.len() {
            return Err(anyhow::anyhow!(
                "no enough bytes to fetch, remaining {} bytes but to fetch {} bytes",
                self.data.len(),
                amt,
            ));
        }
        let bytes = &self.data[..amt];
        self.data = &self.data[amt..];
        Ok(bytes)
    }

    pub fn fetch<T>(&mut self) -> anyhow::Result<&'a T> {
        let len = mem::size_of::<T>();
        let bytes = self.fetch_bytes(len).context(format!(
            "failed to build FUSE request payload type {}",
            std::any::type_name::<T>(),
        ))?;
        let ret = unsafe { (bytes.as_ptr() as *const T).as_ref() };
        match ret {
            Some(obj) => Ok(obj),
            None => Err(anyhow::anyhow!(
                "failed to convert bytes to type={}",
                std::any::type_name::<T>()
            )),
        }
    }
}

fn main(){
    use std::cell::Cell;

    let buf: [u8; 128] = [0u8; 128];

    {
        // Thread A
        let ref_a: &Cell<u128> = ByteSlice::new(&buf).fetch().unwrap();
        ref_a.set(42);
    }
    {
        // Thread B
        let ref_b: &Cell<u128> = ByteSlice::new(&buf).fetch().unwrap();
        ref_b.set(24); // potential race condition here
    }
}
```

Run miri:

```bash
cargo +nightly miri
```

The output:

```
error: Undefined Behavior: accessing memory with alignment 1, but alignment 8 is required
   --> /home/nugine/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/ptr/const_ptr.rs:100:57
    |
100 |         if self.is_null() { None } else { unsafe { Some(&*self) } }
    |                                                         ^^^^^^ accessing memory with alignment 1, but alignment 8 is required
    |
    = help: this usually indicates that your program performed an invalid operation and caused Undefined Behavior
    = help: but alignment errors can also be false positives, see https://github.com/rust-lang/miri/issues/1074
    = help: you can disable the alignment check with `-Zmiri-disable-alignment-check`, but that could hide true bugs
            
    = note: inside `std::ptr::const_ptr::<impl *const std::cell::Cell<u128>>::as_ref` at /home/nugine/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/ptr/const_ptr.rs:100:57
note: inside `ByteSlice::fetch::<std::cell::Cell<u128>>` at src/main.rs:32:28
   --> src/main.rs:32:28
    |
32  |         let ret = unsafe { (bytes.as_ptr() as *const T).as_ref() };
    |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `main` at src/main.rs:50:34
   --> src/main.rs:50:34
    |
50  |         let ref_a: &Cell<u128> = ByteSlice::new(&buf).fetch().unwrap();
    |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: inside closure at /home/nugine/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67:34
    = note: inside closure at /home/nugine/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:52:73
    = note: inside `std::sys_common::backtrace::__rust_begin_short_backtrace::<[closure@std::rt::lang_start_internal::{{closure}}#0::{{closure}}#0 0:&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /home/nugine/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/sys_common/backtrace.rs:130:5
    = note: inside closure at /home/nugine/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:52:13
    = note: inside `std::panicking::r#try::do_call::<[closure@std::rt::lang_start_internal::{{closure}}#0 0:&&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /home/nugine/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:348:40
    = note: inside `std::panicking::r#try::<i32, [closure@std::rt::lang_start_internal::{{closure}}#0 0:&&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe]>` at /home/nugine/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:325:15
    = note: inside `std::panic::catch_unwind::<[closure@std::rt::lang_start_internal::{{closure}}#0 0:&&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /home/nugine/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panic.rs:394:14
    = note: inside `std::rt::lang_start_internal` at /home/nugine/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:51:25
    = note: inside `std::rt::lang_start::<()>` at /home/nugine/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67:5

error: aborting due to previous error

error: could not compile `testrs`.

To learn more, run the command again with --verbose.
```

The unsoundness has three aspects:

1. Unconditional transmutation
2. Unsync access
3. Alignment error (a false positive?)

### PR DESCRIPTION
**Pull request by Nugine**
Fix the unsoundness of `ByteSlice` implementation.

Closes #67.

### CODE REVIEW

File: async_fuse/src/byte_slice.rs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        let strlen: usize = match memchr::memchr(b'\0', self.data) {
+            Some(nul_pos) => nul_pos + 1,
+            None => anyhow::bail!("no trailing zero in bytes, cannot fetch c-string"),
+        };
+        let bytes: &[u8] = self.fetch_bytes(strlen)?;
+        let ret: &CStr = unsafe { CStr::from_bytes_with_nul_unchecked(bytes) };
+        Ok(ret)
+    }
+
+    #[cfg(unix)]
```
Comment (pwang7) [inline async_fuse/src/byte_slice.rs:118]: How about `#[cfg(any((target_os="linux"), (target_os="macos")))]`?
so as to explicitly list supported OS.

File: async_fuse/src/protocol.rs [hunk 2]
```diff
@@ -1117,3 +1117,127 @@ struct FuseCopyFileRangeIn {
     pub len: u64,
     pub flags: u64,
 }
+
+/// FUSE ABI types.
+/// It is safe to transmute a `&[u8]` to `&T` when `T: FuseAbiData`.
+pub(crate) unsafe trait FuseAbiData: Sized {}
+
+macro_rules! unsafe_impl_FuseAbiData_for{
```
Comment (pwang7) [inline async_fuse/src/protocol.rs:1125]: name macro in lowercase, like `unsafe_imple_fuse_abi_data_for`

File: async_fuse/src/byte_slice.rs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn fetch_all_as_slice<T: FuseAbiData>(&mut self) -> anyhow::Result<&'a [T]> {
+        let elem_len: usize = mem::size_of::<T>();
+
+        if self.data.len() % elem_len != 0 {
+            anyhow::bail!(
+                "failed to convert bytes to a slice of type={}, \
+                 the total bytes length={} % the type size={} is nonzero",
```
Comment (pwang7) [inline async_fuse/src/byte_slice.rs:86]: more indentation (4 spaces) for the second line of the string:
```
"failed to convert bytes to a slice of type={}, \
    the total bytes length={} % the type size={} is nonzero",
```

File: async_fuse/src/byte_slice.rs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+                std::any::type_name::<T>(),
+                self.data.len(),
+                elem_len
```
Comment (pwang7) [inline async_fuse/src/byte_slice.rs:89]: elem_len,

File: async_fuse/src/byte_slice.rs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+            );
+        }
+
+        if (self.data.as_ptr() as usize) % mem::align_of::<T>() != 0 {
+            anyhow::bail!(
+                "failed to convert bytes to a slice of type={}, \
+                 pointer={:p} is not a multiple of alignment={}",
+                std::any::type_name::<T>(),
+                self.data.as_ptr(),
+                mem::align_of::<T>()
```
Comment (pwang7) [inline async_fuse/src/byte_slice.rs:99]: mem::align_of::<T>(),

File: async_fuse/src/protocol.rs [hunk 6]
```diff
@@ -1117,3 +1117,127 @@ struct FuseCopyFileRangeIn {
...
+#[cfg(feature = "abi-7-24")]
+unsafe_impl_FuseAbiData_for! {
+    FuseLSeekIn,
+    FuseLSeekOut,
+}
+
+#[cfg(target_os = "macos")]
+unsafe_impl_FuseAbiData_for! {
+    FuseGetXTimesOut,
+    FuseExchangeIn
```
Comment (pwang7) [inline async_fuse/src/protocol.rs:1242]: FuseExchangeIn,

File: async_fuse/src/protocol.rs [hunk 7]
```diff
@@ -1117,3 +1117,127 @@ struct FuseCopyFileRangeIn {
...
+}
+
+#[cfg(feature = "abi-7-21")]
+unsafe_impl_FuseAbiData_for! {
+    FuseDirEntPlus
+}
+
+#[cfg(feature = "abi-7-23")]
+unsafe_impl_FuseAbiData_for! {
+    FuseRename2In
```
Comment (pwang7) [inline async_fuse/src/protocol.rs:1230]: FuseRename2In,

File: async_fuse/src/protocol.rs [hunk 8]
```diff
@@ -1117,3 +1117,127 @@ struct FuseCopyFileRangeIn {
...
+
+#[cfg(feature = "abi-7-19")]
+unsafe_impl_FuseAbiData_for! {
+    FuseFAllocateIn,
+
+}
+
+#[cfg(feature = "abi-7-21")]
+unsafe_impl_FuseAbiData_for! {
+    FuseDirEntPlus
```
Comment (pwang7) [inline async_fuse/src/protocol.rs:1225]: FuseDirEntPlus,

File: async_fuse/src/protocol.rs [hunk 9]
```diff
@@ -1117,3 +1117,127 @@ struct FuseCopyFileRangeIn {
...
+    FuseNotifyInvalEntryOut,
+    FuseNotifyInvalINodeOut,
+}
+
+#[cfg(feature = "abi-7-15")]
+unsafe_impl_FuseAbiData_for! {
+    FuseNotifyRetrieveOut,
+    FuseNotifyRetrieveIn,
+    FuseNotifyStoreOut,
+    FuseCopyFileRangeIn
```
Comment (pwang7) [inline async_fuse/src/protocol.rs:1201]: FuseCopyFileRangeIn,

File: async_fuse/src/protocol.rs [hunk 10]
```diff
@@ -1117,3 +1117,127 @@ struct FuseCopyFileRangeIn {
...
+    FuseLockOut,
+    FuseAccessIn,
+    FuseInitIn,
+    FuseInitOut,
+    FuseInterruptIn,
+    FuseBMapIn,
+    FuseBMapOut,
+    FuseInHeader,
+    FuseOutHeader,
+    FuseDirEnt
```
Comment (pwang7) [inline async_fuse/src/protocol.rs:1171]: FuseDirEnt,

File: async_fuse/src/protocol.rs [hunk 11]
```diff
@@ -1117,3 +1117,127 @@ struct FuseCopyFileRangeIn {
...
+
+macro_rules! unsafe_impl_FuseAbiData_for{
+    {$($t:ty),+} => {
+        $(unsafe impl FuseAbiData for $t {})+
+    }
+}
+
+unsafe_impl_FuseAbiData_for! {
+    u8, u16, u32, u64, usize,
+    i8, i16, i32, i64, isize
```
Comment (pwang7) [inline async_fuse/src/protocol.rs:1133]: do we need this?

File: async_fuse/src/fuse_request.rs [hunk 12]
```diff
@@ -725,67 +646,81 @@ impl<'a> Request<'a> {
 #[cfg(test)]
 mod test {
     use super::*;
-    use log::debug;
+
+    #[repr(align(8))]
```
Comment (pwang7) [inline async_fuse/src/fuse_request.rs:650]: why `repr(align(8))`? is there any special consideration?

File: async_fuse/src/protocol.rs (hunk 11 above)
Comment (Nugine) [reply async_fuse/src/protocol.rs:1133]: we need this impl for struct fields

File: async_fuse/src/fuse_request.rs (hunk 12 above)
Comment (Nugine) [reply async_fuse/src/fuse_request.rs:650]: `FuseInHeader` is aligned to 8 bytes. `Align8` is enough for structs used here.
`[u8;N]` is aligned to 1 byte.  Requests which are not well-aligned will cause an alignment error (potential UB).
So we have runtime checks in `ByteSlice::fetch` and `ByteSlice::fetch_all_as_slice`.

Described in <https://github.com/datenlord/datenlord/pull/84/commits/bb4ea0b224b3ccbccb1645ca1dbe1a3dbd24ed87>

File: async_fuse/src/protocol.rs (hunk 10 above)
Comment (Nugine) [reply async_fuse/src/protocol.rs]: done

File: async_fuse/src/protocol.rs (hunk 7 above)
Comment (Nugine) [reply async_fuse/src/protocol.rs]: done

File: async_fuse/src/protocol.rs (hunk 6 above)
Comment (Nugine) [reply async_fuse/src/protocol.rs]: done

File: async_fuse/src/byte_slice.rs (hunk 5 above)
Comment (Nugine) [reply async_fuse/src/byte_slice.rs]: done

File: async_fuse/src/byte_slice.rs (hunk 4 above)
Comment (Nugine) [reply async_fuse/src/byte_slice.rs]: done

File: async_fuse/src/protocol.rs (hunk 2 above)
Comment (Nugine) [reply async_fuse/src/protocol.rs]: done

File: async_fuse/src/byte_slice.rs (hunk 1 above)
Comment (Nugine) [reply async_fuse/src/byte_slice.rs]: done

File: async_fuse/src/byte_slice.rs (hunk 3 above)
Comment (Nugine) [reply async_fuse/src/byte_slice.rs]: done

File: async_fuse/src/protocol.rs (hunk 8 above)
Comment (Nugine) [reply async_fuse/src/protocol.rs]: done

File: async_fuse/src/protocol.rs (hunk 9 above)
Comment (Nugine) [reply async_fuse/src/protocol.rs]: done

**Comment by Nugine** [comment]
ready to merge

**Comment by Nugine** [comment]
ready to merge

This PR only fixes a part of the unsafe problem. Other unsafe code has not been verified yet.

File: async_fuse/src/fuse_request.rs [hunk 13]
```diff
@@ -731,67 +650,88 @@ impl<'a> Request<'a> {
...
+    struct Align8<T: Sized>(T);
```
Comment (pwang7) [inline async_fuse/src/fuse_request.rs:655]: should use AlignedBytes here?

**Pull request merged**